contracts/src/Zappers/GasCompZapper.sol

```diff
@@ -261,5 +262,67 @@ contract GasCompZapper is AddRemoveManagers, LeftoversSweep {
         require(success, "GCZ: Sending ETH failed");
     }
 
+    function closeTroveFromCollateral(CloseTroveParams memory _params) external override {
+        address owner = troveNFT.ownerOf(_params.troveId);
+        address payable receiver = payable(_requireSenderIsOwnerOrRemoveManagerAndGetReceiver(_params.troveId, owner));
+        _params.receiver = receiver;
+
+        // Set initial balances to make sure there are not lefovers
+        InitialBalances memory initialBalances;
+        _setInitialBalances(collToken, boldToken, initialBalances);
+
+        // Flash loan coll
+        flashLoanProvider.makeFlashLoan(
+            collToken, _params.flashLoanAmount, IFlashLoanProvider.Operation.CloseTrove, abi.encode(_params)
+        );
+
+        // return leftovers to user
+        _returnLeftovers(collToken, boldToken, initialBalances);
+    }
+
+    function receiveFlashLoanOnCloseTroveFromCollateral(
+        CloseTroveParams calldata _params,
+        uint256 _effectiveFlashLoanAmount
+    ) external {
+        require(msg.sender == address(flashLoanProvider), "GCZ: Caller not FlashLoan provider");
+
+        LatestTroveData memory trove = troveManager.getLatestTroveData(_params.troveId);
+
+        // Swap Coll from flash loan to Bold, so we can repay and close trove
+        // We swap the flash loan minus the flash loan fee
+        exchange.swapToBold(_effectiveFlashLoanAmount, trove.entireDebt);
+
+        // We asked for a min of entireDebt in swapToBold call above, so we don’t check again here:
+        //uint256 receivedBoldAmount = exchange.swapToBold(_effectiveFlashLoanAmount, trove.entireDebt);
+        //require(receivedBoldAmount >= trove.entireDebt, "GCZ: Not enough BOLD obtained to repay");
+
+        borrowerOperations.closeTrove(_params.troveId);
+
+        // Send coll back to return flash loan
+        collToken.safeTransfer(address(flashLoanProvider), _params.flashLoanAmount);
+
+        // Send coll left
+        collToken.safeTransfer(_params.receiver, trove.entireColl - _params.flashLoanAmount);
```

Comment:
Hmm, there's a receiver parameter, but leftovers are still swept to `msg.sender`, is that intended?

Reply:
Also, should we add an early check for `trove.entireColl >= _params.flashLoanAmount` to avoid panicking so late?

Reply:
I considered that, but in this case, being the reason a user provided param, I think it’s “user fault” (probably more the frontend but still :man_shrugging: ), so no big deal if some more gas is spent. Otherwise, we are making “good users” spend more unnecessary gas with an extra check.
Or even without the bad/good judgement: statistically how many times the happy path would happen vs the one where that param is wrong? I guess it wouldn’t pay off.

Reply:
> Hmm, there's a receiver parameter, but leftovers are still swept to msg.sender, is that intended?

In the generic case, it’s intended, as it’s `msg.sender` who sends the original funds, but I guess it doesn’t make sense in this case, as funds are not being sent.